kroxylicious-integration-tests/src/test/java/io/kroxylicious/proxy/TlsIT.java

```diff
@@ -162,6 +171,88 @@ void upstreamUsesTlsInsecure(@Tls KafkaCluster cluster) {
         }
     }
 
+    @Test
+    void upstreamRequiresTlsClientAuth_TrustStore() throws Exception {
+        // Note that the annotation driven Kroxylicious Extension doesn't support configuring a cluster that expects client-auth.
+
+        var brokerCert = new KeytoolCertificateGenerator();
+        var clientCert = new KeytoolCertificateGenerator();
+        clientCert.generateSelfSignedCertificateEntry("clientTest@kroxylicious.io", "client", "Dev", "Kroxylicious.ip", null, null, "US");
+
+        try (var cluster = KafkaClusterFactory.create(KafkaClusterConfig.builder()
+                .brokerKeytoolCertificateGenerator(brokerCert)
+                // note passing client generator causes ssl.client.auth to be set 'required'
+                .clientKeytoolCertificateGenerator(clientCert)
+                .kraftMode(true)
+                .securityProtocol("SSL")
+                .build())) {
+            cluster.start();
+
+            var config = new HashMap<>(cluster.getKafkaClientConfiguration());
+
+            // Get the client key material provided by the framework. This will be used to configure Kroxylicious.
+            var keyStore = config.get("ssl.keystore.location").toString();
+            var keyPassword = config.get("ssl.keystore.password").toString();
+            assertThat(keyStore).isNotNull();
+            assertThat(keyPassword).isNotNull();
+
+            var trustStore = config.get("ssl.truststore.location").toString();
+            var trustPassword = config.get("ssl.truststore.password").toString();
+            assertThat(trustStore).isNotNull();
+            assertThat(trustPassword).isNotNull();
+
+            // Validate a TLS client-auth connection to direct the cluster succeeds/fails as expected.
+            assertSuccessfulDirectClientAuthConnectionWithClientCert(cluster);
+            assertUnsuccessfulClientAuthDirectConnectionWithoutClientCert(cluster);
+
+            var builder = new ConfigurationBuilder()
```

Comment:
Maybe the rest of the code should be wrapped up in a method as above like `assertKroxyliciousCanProxyConnectionWithClientCert(...)`. Would it be worth also doing the matching negative test that another Kroxy instance cannot proxy the connection if it's certificates aren't in order?

Reply:
Thanks. I tried the suggestion but in the end backed it out. The issue is that whilst the admin connection via the proxy fails, but there's no way for the upstream error why to be communicated back to the downstream.  The test can only assert a failed connection, which doesn't seem to compelling.

 We really need something (exposed through a management layer in Kroxylicious) that reveals the fact that the upstream connection is failing and why.